Untrusted Validator Squelching - Extends  squelching to limit untrusted validator message propagation

Tapanito opened the pull request:
This feature improves network efficiency by limiting message propagation from untrusted validators.

Squelching currently reduces the volume of duplicate messages from validators but does not address the volume of unique messages from untrusted validators, who may not contribute meaningfully to network progress.

This change introduces a bounded number of slots for untrusted validators, selected based on unique message frequency and quantity. Once selected, their duplicate messages are subject to standard squelching logic, thereby reducing overall message overhead without impacting trusted validator performance.

<!--
This PR template helps you to write a good pull request description.
Please feel free to include additional useful information even beyond what is requested below.

If your branch is on a personal fork and has a name that allows it to
run CI build/test jobs (e.g. "ci/foo"), remember to rename it BEFORE
opening the PR.  This avoids unnecessary redundant test runs. Renaming
the branch after opening the PR will close the PR.
https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-branches-in-your-repository/renaming-a-branch
-->

## High Level Overview of Change

<!--
Please include a summary of the changes.
This may be a direct input to the release notes.
If too broad, please consider splitting into multiple PRs.
If a relevant task or issue, please link it here.
-->

### Context of Change

<!--
Please include the context of a change.
If a bug fix, when was the bug introduced? What was the behavior?
If a new feature, why was this architecture chosen? What were the alternatives?
If a refactor, how is this better than the previous implementation?

If there is a spec or design document for this feature, please link it here.
-->

### Type of Change

<!--
Please check [x] relevant options, delete irrelevant ones.
-->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Performance (increase or change in throughput and/or latency)
- [ ] Tests (you added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation update
- [ ] Chore (no impact to binary, e.g. `.gitignore`, formatting, dropping support for older tooling)
- [ ] Release

### API Impact

<!--
Please check [x] relevant options, delete irrelevant ones.

* If there is any impact to the public API methods (HTTP / WebSocket), please update https://github.com/xrplf/rippled/blob/develop/API-CHANGELOG.md
  * Update API-CHANGELOG.md and add the change directly in this PR by pushing to your PR branch.
* libxrpl: See https://github.com/XRPLF/rippled/blob/develop/docs/build/depend.md
* Peer Protocol: See https://xrpl.org/peer-protocol.html
-->

- [ ] Public API: New feature (new methods and/or new fields)
- [ ] Public API: Breaking change (in general, breaking changes should only impact the next api_version)
- [ ] `libxrpl` change (any change that may affect `libxrpl` or dependents of `libxrpl`)
- [x] Peer protocol change (must be backward compatible or bump the peer protocol version)

<!--
## Before / After
If relevant, use this section for an English description of the change at a technical level.
If this change affects an API, examples should be included here.

For performance-impacting changes, please provide these details:
1. Is this a new feature, bug fix, or improvement to existing functionality?
2. What behavior/functionality does the change impact?
3. In what processing can the impact be measured? Be as specific as possible - e.g. RPC client call, payment transaction that involves LOB, AMM, caching, DB operations, etc.
4. Does this change affect concurrent processing - e.g. does it involve acquiring locks, multi-threaded processing, or async processing?
-->

<!--
## Test Plan
If helpful, please describe the tests that you ran to verify your changes and provide instructions so that others can reproduce.
This section may not be needed if your change includes thoroughly commented unit tests.
-->

<!--
## Future Tasks
For future tasks related to PR.
-->

Tapanito committed:
adds config option for enhanced squelching

Tapanito committed:
adds isTrusted parameter to updateSlotAndSquelch method to differentiate messages from trusted validators

Tapanito committed:
adds methods to track which peers and validators were squelched

Tapanito committed:
adds method to SquelchHandler to squelch all peers

Tapanito committed:
adds data strcutures and methods to track red validators

Tapanito committed:
feature: extend squelching to suppress untrusted validator traffic
This feature improves network efficiency by limiting message propagation from untrusted validators.

Squelching currently reduces the volume of duplicate messages from validators but does not address the volume of unique messages from untrusted validators, who may not contribute meaningfully to network progress.

This change introduces a bounded number of slots for untrusted validators, selected based on message frequency. Once selected, their duplicate messages are subject to standard squelching logic, thereby reducing overall message overhead without impacting trusted validator performance.

Tapanito committed:
adds methods to print slot details from  command

Tapanito committed:
adds enhanced squelching tests

Tapanito committed:
removes member functions from Slots used only for testing, and moves them to tests

Tapanito committed:
fixes unittests for windows

vlntb commented on src/xrpld/overlay/detail/OverlayImpl.cpp:
```diff
@@ -578,16 +578,23 @@ OverlayImpl::stop()
-    beast::PropertyStream::Set set("traffic", stream);
-    auto const stats = m_traffic.getCounts();
-    for (auto const& pair : stats)
     {
```
I can see the intention to destroy beast::PropertyStream::Set before sending the stream object into onWrite, but I don't understand why is this necessary.

Tapanito replied:
To have separate categories in a single map, we first need to update one category, flush it, then work on the next category. The flushing logic is called in the deconstructor (see: [here](https://github.com/XRPLF/rippled/blob/develop/src/libxrpl/beast/utility/beast_PropertyStream.cpp)). 

So, if our map has two elements, A and B. We first have to finish with A, flush it, then work with B. Here is an example where it is done this way elsewhere in the code: https://github.com/XRPLF/rippled/blob/develop/include/xrpl/resource/detail/Logic.h .

Tapanito commented on src/test/overlay/clock.h:
```diff
@@ -0,0 +1,83 @@
+using namespace std::chrono;
+
+/** Manually advanced clock. */
+class ManualClock
```
This call was moved from the original tests that Greg wrote, so that it could be reused inside enhanced squelching tests.

I did review the manual_clock class before. It is an instanced class, meaning that it would have to be passed as an argument to the constructor of `Slots` . While I agree that this would be better, such refactoring is beyond the scope of this feature.

Tapanito replied:
I removed the ManualClock, which required refactoring the entire base and enhanced squelching implementation :) 
The reason being is that the old code used templating to pass ManualClock for tests. The moment I removed ManualClock it no longer made sense to use templates, so I removed them.  Removing templates forced to split the old self contained header file (Slot.h and Squelch.h) into Slot.[h,cpp] and Squelch.[h,cpp].

Tapanito committed:
Update src/xrpld/overlay/detail/OverlayImpl.h

Co-authored-by: Valentin Balaschenko <13349202+vlntb@users.noreply.github.com>

Tapanito committed:
removes duplicate untrusted slot check

Tapanito commented:
@bthomee, it would require rewriting a good chunk of existing base
squelching tests, I think that's out of scope. However, I'm happy to
explore this prospect separately.

On Mon, Jun 23, 2025, 20:13 Bart ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/test/overlay/clock.h
> <https://github.com/XRPLF/rippled/pull/5399#discussion_r2162216340>:
>
> > +#define RIPPLE_TEST_OVERLAY_CLOCK_H_INCLUDED
> +
> +#include <xrpl/basics/random.h>
> +
> +#include <chrono>
> +#include <cstdint>
> +#include <ratio>
> +
> +namespace ripple {
> +
> +namespace test {
> +
> +using namespace std::chrono;
> +
> +/** Manually advanced clock. */
> +class ManualClock
>
> @Tapanito <https://github.com/Tapanito> Adding an extra class that
> duplicates a lot of existing code/functionality leads to code bloat.
> Passing an instance around shouldn't be a huge refactor.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/XRPLF/rippled/pull/5399#discussion_r2162216340>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABMDKU2YWCR3NW4WIEAH2R33FA7VJAVCNFSM6AAAAAB25WMZ7CVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDSNJRGA2DAMZZGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

Tapanito committed:
refactors squelching to use instanced clock instead of a static clock

Tapanito committed:
improves code readabiliy

Tapanito committed:
removes unused clock

Tapanito committed:
removes unused imports

Tapanito committed:
removes unused test parameters

gregtatcam commented on src/xrpld/overlay/Slot.h:
```diff
@@ -257,333 +267,60 @@ class Slot final
 {
-    using time_point = typename clock_type::time_point;
-    using id_t = typename Peer::id_t;
+    friend test::enhanced_squelch_test;
```
Same as above. Decouple from the unit-tests.

gregtatcam commented on src/xrpld/overlay/Slot.h:
```diff
@@ -593,244 +330,181 @@ class Slots final
-    hash_map<PublicKey, Slot<clock_type>> slots_;
-    SquelchHandler const& handler_;  // squelch/unsquelch handler
+    slots_map slots_;
+    slots_map untrusted_slots_;
```
We use Hungarian notation for the data members.

Tapanito replied:
Done

gregtatcam commented on src/xrpld/overlay/detail/Slot.cpp:
```diff
@@ -0,0 +1,653 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012, 2013 Ripple Labs Inc.
```
Should update the copyright.

gregtatcam commented on src/xrpld/overlay/detail/Squelch.cpp:
```diff
@@ -0,0 +1,79 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2020 Ripple Labs Inc.
```
Should update the copyright.

gregtatcam commented on src/xrpld/overlay/Slot.h:
```diff
@@ -593,244 +330,181 @@ class Slots final
+     * @return true if a given validator was squelched
+     */
+    bool
+    validatorSquelched(PublicKey const& validatorKey);
```
This and next one should be split into two functions. One is public const to return the state and another expire and return the state. Otherwise they could be used to check for the state and unintentionally expire. Perhaps something like `expireAndIsValidatorSquelched`.

Tapanito replied:
This is intended behaviour. If a squelch for a validator or a peer has expired, it should no longer be squelched, thus expiring old entries is always a desired behaviour.  Separating this logic into two separate methods would mean that a validator or a peer whose squelch has expired would wrongfully be marked as squelched, where in reality they're expired.

Tapanito committed:
improves code readability

Tapanito committed:
decouples tests from squelching implementation

Tapanito commented on src/xrpld/overlay/Slot.h:
```diff
@@ -22,28 +22,30 @@
 
 namespace ripple {
+// used to make private members of Slots class accessible for testing
+namespace test {
```
Done

Tapanito commented on src/xrpld/overlay/Slot.h:
```diff
@@ -101,13 +135,11 @@ class SquelchHandler
-    friend class Slots<clock_type>;
-    using id_t = Peer::id_t;
-    using time_point = typename clock_type::time_point;
+    friend class Slots;
```
Done

Tapanito commented on src/xrpld/overlay/Slot.h:
```diff
@@ -22,28 +22,30 @@
 namespace reduce_relay {
 
-template <typename clock_type>
+using clock_type = beast::abstract_clock<std::chrono::steady_clock>;
```
Done

Tapanito commented on src/xrpld/overlay/Slot.h:
```diff
@@ -22,28 +22,30 @@
-template <typename clock_type>
+using clock_type = beast::abstract_clock<std::chrono::steady_clock>;
+using time_point = clock_type::time_point;
+
```
Done

Tapanito committed:
fixes code formatting

Tapanito committed:
fixes windows tests

gregtatcam commented on src/xrpld/overlay/Slot.h:
```diff
@@ -593,244 +338,182 @@ class Slots final
+    // Maintain aged container of validator/peers. This is used to track
+    // which validator/peer were squelced. A peer that whose squelch
+    // has expired is removed.
+    validators peersWithValidators_;
```
Is the info you get from this container different from what you can get from `slots_`? I.e. `slots` is `pubKey: Slot`, where `Slot` has a list of peers: `peerId:PeerInfo`, where `PeerInfo` has the peer state. So this is basically the same as `validators`.

Tapanito replied:
I'll change the name, its good!

This is a map from validator to a set of peers, indicating which peers were sent a squelch message for the particular validator. It is different from `slots_` and `untrustedSlots_`. Once `untrustedSlots_` are full, additional validators will not be added to them, but new untrusted validators will be squelched, and this squelching is tracked in `peersWithSquelchedValidators_`.

gregtatcam commented on src/xrpld/overlay/detail/Slot.cpp:
```diff
@@ -0,0 +1,655 @@
+    if (peer.count == (MAX_MESSAGE_THRESHOLD + 1))
+        ++reachedThreshold_;
+
+    if (now - lastSelected_ > 2 * MAX_UNSQUELCH_EXPIRE_DEFAULT)
```
This is defensive block of code. But if it ever happens then we stuck in this state since `lastSelected_` is updated once the peers are selected below. This should never really happen since a slot is deleted if `now - lastSelected_ > MAX_UNSQUELCH_EXPIRE_DEFAULT)`. My question is if `lastSelected_` should be set to `now` if above is true or should we make this an error and delete this slot?

gregtatcam replied:
So basically the time of the most recent message? I think we just should delete the slot in this case since. This scenario can't happen.

Tapanito replied:
I think it can happen. Imagine we have a slot for untrusted validator that stops sending traffic. We will very quickly learn that it's offline, but it'll take an hour until the slot is deleted. Though, and I already started working on this, for untrusted validators, if they fall under the threshold of minimum peers, they slot will be deleted and the validator squelched.

gregtatcam commented on src/xrpld/overlay/detail/OverlayImpl.cpp:
```diff
@@ -578,16 +578,23 @@ OverlayImpl::stop()
+        }
+    }
+
+    {
```
`set` is destroyed in the first block. There is nothing that goes out of scope in this block, is it?

Tapanito committed:
improves code readability

Tapanito commented on src/xrpld/overlay/detail/Slot.cpp:
```diff
@@ -0,0 +1,653 @@
+    beast::expire(peersWithMessage_, reduce_relay::IDLED);
+
+    if (key.isNonZero())
+    {
```
It can be, I purposefully haven't touched the original squelching code. There's a fair bit of refactoring that can be done around it, I'm not sure whether it should be done with this PR, or separetely. What do you think?

Tapanito commented on src/xrpld/overlay/Slot.h:
```diff
@@ -593,244 +338,185 @@ class Slots final
 
-    hash_map<PublicKey, Slot<clock_type>> slots_;
-    SquelchHandler const& handler_;  // squelch/unsquelch handler
+    slots_map slots_;
```
That's right. However, on a second though, I think it is confusing to stored untrusted validators in slots_ if enhanced squelching is disabled! I've updated the code to always store untrusted validators in untrustedSlots_ and trusted validators in slots_. Which in turn, let's me to rename slots_ to `trustedSlots_`

Tapanito commented on src/xrpld/overlay/detail/Slot.cpp:
```diff
@@ -0,0 +1,653 @@
+    if (now - it->second.lastMessage > IDLED)
+        return {};
+
+    it->second.peers.insert(peer);
```
`peers` is a set, I will not know it's true size until the value is insert in line `522`, thus just adding one in the `if (...)`  check is not right.  While I understand where you're coming from regarding the redundant update in case we delete the entry, I don't think it's such a computationally heavy operation.

gregtatcam replied:
Ah, right. It could be a duplicate peer.

Tapanito committed:
further improves member function and attribute names

Tapanito committed:
adds callback to squelchAll instead of calling slots directly

Tapanito commented on src/xrpld/overlay/Slot.h:
```diff
@@ -75,14 +88,23 @@ class SquelchHandler
+     * @param duration Squelch duration in seconds
+     */
+    virtual void
+    squelchAll(PublicKey const& validator, std::uint32_t duration) = 0;
```
Done!

Tapanito committed:
removes redundant scope

Tapanito committed:
adds logic to reset validator progress and better deletion safeguards

If the validator was idle for a short period, reset it's progress. However, if the validator was idle for a long time, delete and squelch it.
Similarly, if the validator sent a lot of unique messages, but failed to reach peering constraints, squelch it.

Tapanito committed:
improves code readability

Tapanito committed:
changes selection parameters and number of untrusted slots

Simulation results revealed that by requiring a minimum number of peers to send a message, valdiations will stop propagating, as the servers that are directly connected to peers will receive a unique message from only a single peer, the validator. Therefore, they will fail to select a slot.
Furthermore, by increasing untrusted slots to 30 we guarantee that some validators will propagate through most of the network.

Tapanito committed:
adds squelchStore

Tapanito committed:
extends enhanced squelching to squelch a validator squelched by majority of peers

Tapanito commented on src/xrpld/overlay/detail/OverlayImpl.cpp:
```diff
@@ -1410,49 +1420,115 @@ OverlayImpl::squelch(
+OverlayImpl::squelchAll(
+    PublicKey const& validator,
+    uint32_t squelchDuration,
+    std::function<void(Peer::id_t)> callback)
```
`register` is un-used reserved keyword: https://en.cppreference.com/w/cpp/keyword/register.html

Tapanito committed:
removes redundant checks before erasing elements

Tapanito committed:
adds documentation and improves squelching comments

Tapanito committed:
improves code readability

vlntb commented on src/xrpld/overlay/SquelchStore.h:
```diff
@@ -0,0 +1,144 @@
+     * @brief The core data structure mapping a validator's public key to the
+     * time point when their squelch expires.
+     */
+    hash_map<PublicKey, time_point> squelched_;
```
Couple of questions:

1. What is the guarantee that the validator will disappeare and will be removed from the list, but `squelched_` will still keep an entry for it? Can we have non-deleted dead entries in this hash_map?

2. How big can this hash_map get? In some case for a small number of pairs, the other collection might be more efficient. I'm assuming that this data structure is on the hot path.

Tapanito replied:
> What is the guarantee that the validator will disappeare and will be removed from the list, but squelched_ will still keep an entry for it? Can we have non-deleted dead entries in this hash_map?


Entries are deleted when `expiredAndIsSquelched` is called. In turn, this method is called whenever a server needs to decide whether to forward a message or not. Consequentially, if a validator goes offline, then the server will not check whether a squelch for it expired, and the entry may stay in the map until the `PeerImpl` instance is destroyed!

The `PeerImpl` already has a timer task, we could add a call there to clean up dead entries. However, the limitation with this approach is that then we would have to introduce a mutex into `SquelchStore`. However, this is not ideal, as it would place a mutex on critical path of deciding whether to relay a validator message.




> How big can this hash_map get? In some case for a small number of pairs, the other collection might be more efficient. I'm assuming that this data structure is on the hot path.

In the worse case, the `SquelchStore` will have as many entries as there are validators in the network. However, technically, a peer could spam us with non-existent validator keys, and thus explode the memory cap. We could, however, add an artificial limit, of let's say, 500 entries to safeguard from this. Given that a hash_map provides O(1) lookup, what other data structure would be better?

vlntb replied:
1. Yes, we need some way to detect those orphaned validator entries. 
How about modified the logic below to also include clearing them?


```
void
OverlayImpl::handleUntrustedSquelch(PublicKey const& validator)
{
    if (!strand_.running_in_this_thread())
        return post(
            strand_,
            std::bind(&OverlayImpl::handleUntrustedSquelch, this, validator));

    auto count = 0;
    // we can get the total number of peers with size(), however that would have
    // to acquire another lock on peers. Instead, count the number of peers in
    // the same loop, as we're already iterating all peers.
    auto total = 0;
    for_each([&](std::shared_ptr<PeerImp>&& p) {
        ++total;
        if (p->expireAndIsSquelched(validator))
            ++count;
    });

    // if majority of peers squelched the validator
    if (count >= total - 1)
        slots_.squelchUntrustedValidator(validator);
}
```

vlntb commented on src/test/overlay/enhanced_squelch_test.cpp:
```diff
@@ -0,0 +1,981 @@
+    }
+};
+
+class EnhancedSquelchingTestSlots : public reduce_relay::Slots
```
This is a very comprehensive test suite!

vlntb commented on src/xrpld/overlay/Slot.h:
```diff
@@ -177,660 +283,620 @@ class Slot final
-                << "deleteIdlePeers: deleting idle slot " << Slice(it->first);
-            it = slots_.erase(it);
+            count = 0;
+            peers.clear();
```
```suggestion
            peers.clear();
            lastMessage = time_point{};
```

vlntb commented on src/xrpld/overlay/detail/PeerImp.cpp:
```diff
@@ -43,7 +44,10 @@
 #include <boost/algorithm/string/predicate.hpp>
 #include <boost/beast/core/ostream.hpp>
 
+#include "xrpld/overlay/Peer.h"
```
The convention within rippled is to use angle-bracket includes.

vlntb commented on src/test/overlay/enhanced_squelch_test.cpp:
```diff
@@ -0,0 +1,981 @@
+#include <xrpl/beast/unit_test.h>
+#include <xrpl/protocol/SecretKey.h>
+
+#include "xrpld/overlay/Peer.h"
```
The convention within rippled is to use angle-bracket includes.

Tapanito committed:
improves expired squelch deletion

Tapanito committed:
Merge branch 'develop' into tapanito/feature/enhanced-squelching

gregtatcam commented on src/xrpld/overlay/Slot.h:
```diff
@@ -593,244 +338,182 @@ class Slots final
-            ++it;
-    }
-}
+    hash_map<PublicKey, ValidatorInfo> consideredValidators_;
```
A reminder to consider renaming ^^. After all there are now `trustedSlots_` and `untrustedSlots_`. It makes sense to be clear what this variable refers to. It makes the code more readable. This also applies to the functions changing the state of this variable.

gregtatcam commented on src/xrpld/overlay/detail/SquelchStore.cpp:
```diff
@@ -0,0 +1,103 @@
+
+    if (squelch)
+    {
+        // This should never trigger. The squelh duration is validated in
```
Typo: `squelh` -> `squelch`.

Tapanito commented on src/xrpld/overlay/detail/PeerImp.cpp:
```diff
@@ -29,6 +29,8 @@
 #include <xrpld/app/misc/ValidatorList.h>
 #include <xrpld/app/tx/apply.h>
 #include <xrpld/overlay/Cluster.h>
+#include <xrpld/overlay/Peer.h>
```
Not challenging, trying to understand: Why not? We are using `Peer::id_t`, isn't it better to explicitly include it, rather than relying on transitive includes?

Tapanito replied:
Removed, as `PeerImpl` directly inherits from `Peer`

Tapanito commented on src/xrpld/overlay/detail/OverlayImpl.cpp:
```diff
@@ -24,8 +24,11 @@
 #include <xrpld/overlay/detail/PeerImp.h>
+#include <xrpld/overlay/detail/TrafficCount.h>
 #include <xrpld/overlay/detail/Tuning.h>
 #include <xrpld/overlay/predicates.h>
```
Why not?  OverlayImpl uses types from TrafficCount

Tapanito commented on src/xrpld/overlay/detail/PeerImp.cpp:
```diff
@@ -2395,6 +2389,23 @@ PeerImp::onMessage(std::shared_ptr<protocol::TMValidation> const& m)
+            // dropped then this happens here I.e. before further wasting CPU
+            // verifying the signature of an untrusted key
+            if (app_.config().RELAY_UNTRUSTED_VALIDATIONS == -1)
+                return;
```
That's a great point @gregtatcam .

I think, once enhanced squelching is deployed widely, we should deprecate this configuration option, as it limits the relaying of all untrusted validators, not just the select few. I'll update the comment to reflect this.

Tapanito committed:
clean up imports, and typos

Tapanito committed:
ammends documentation changig callback to report

gregtatcam commented on src/xrpld/overlay/detail/OverlayImpl.h:
```diff
@@ -24,6 +24,7 @@
 #include <xrpld/core/Job.h>
 #include <xrpld/overlay/Message.h>
 #include <xrpld/overlay/Overlay.h>
+#include <xrpld/overlay/Peer.h>
```
It's not that you don't need this include, it's just it's not used because it's already included in `Overlay.h`.

gregtatcam commented on src/xrpld/overlay/detail/PeerImp.cpp:
```diff
@@ -44,6 +46,7 @@
 #include <boost/beast/core/ostream.hpp>
 
 #include <algorithm>
+#include <chrono>
```
It's not used. It's already included in one of the other includes.

gregtatcam reviewed (dismissed):
LGTM

Tapanito committed:
Merge branch 'develop' into tapanito/feature/enhanced-squelching

Tapanito committed:
forces a copy of reference parameter in OverlayImpl::updateUntrustedValidatorSlot

ximinez commented:
@Tapanito I notice @mvadari tagged this as needing an XLS. Is that in progress? I assume we'd want to have that befoe this is merged.

Tapanito committed:
Merge branch 'develop' into tapanito/feature/enhanced-squelching

Tapanito committed:
Merge branch 'develop' into tapanito/feature/enhanced-squelching

Tapanito committed:
clang-tidy

Tapanito committed:
clang-tidy
